signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+    ) -> None:
+        """
+        Sets exporter object fields and generates unique publisher_id.
+        Allows for specification of redis host/port. Also allows runner
+        hostname to be inputted manually (otherwise will default to 
+        platform.node() value)
+        """
+        self.subs = []
+        self.proc_name = process_name
+        self.runner_host = runner_host
+        self.pub_id = process_name + "-" + str(uuid.uuid4())
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.init_listener = None
+        self.legal_events = None
+
+    def _sig_builder(
+        self, event: str, sample: int = -1, tag: str = None, metadata: Dict = None
+    ) -> Signal:
+        """
+        Build a signal data object based on exporter object fields,
+        as well as user-inputted fields. Returns the signal object.
+        """
+        sig = Signal(publisher_id=self.pub_id, process_name=self.proc_name, event=event)
+        sig.runner_host = self.runner_host if self.runner_host else sig.runner_host
+        sig.tag = tag if tag else sig.tag
+        sig.metadata = metadata if metadata else sig.metadata
+        sig.sample_no = sample if sample >= 0 else sig.sample_no
+        return sig
+
+    def _get_data_dict(self, response: Dict) -> Dict:
+        """
+        Returns response signal payload if a properly-formed
+        response is received. Otherwise return None.
+        """
+        if not "data" in response:
+            print("No data in this response message")
+            return None
+        try:
+            data = json.loads(response["data"])
+        except ValueError:
+            return None
+        if "tool_id" not in data or "publisher_id" not in data or "event" not in data:
+            print("Malformed response data found")
+            return None
+        return data
+
+    def _fetch_responders(self) -> None:
+        """
+        Start initialization response listener. Add tool_ids from proper
+        responses to the subscriber list.
+        """
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+
+        def _init_handler(item) -> None:
+            data = self._get_data_dict(item)
+            if (
+                data
+                and data["event"] == "initialization"
+                and data["publisher_id"] == self.pub_id
+            ):
+                self.subs.append(data["tool_id"])
+
+        subscriber.subscribe(**{"event-signal-response": _init_handler})
+        self.init_listener = subscriber.run_in_thread()
+
+    def _check_subs(self, event: str) -> Tuple[Any, list]:
+        """
+        Listen for responses from all registered subscribers. Return
+        listener, as well as value based on responders' RAS codes.
+        """
+        if not self.subs:
+            return None, [0]
+
+        to_check = set(self.subs)
+        subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        result_box = [0]
```

Comment:
Are you using an array here to get around issues with scope?

Reply:
Yeah, it's so that the listener can return a result code value that can be changed depending on responses

Reply:
Ah I see ok. If only Python had pointers right? My only trouble with this is that it is a bit hard to read for those not familiar with the intricacies of Python's atomic built-ins and the GIL that allow for a list like this to essentially work like a queue- I had to do quite a bit of Googling to understand how this worked. I know that this does work, and it's so cool that it does, but IMHO I would prefer if we used the built-in `queue` module. See Google's Python style guide: https://google.github.io/styleguide/pyguide.html#218-threading